xbmc/cores/VideoPlayer/DVDInputStreams/InputStreamAddon.cpp

```diff
@@ -20,40 +20,143 @@
 
 #include "InputStreamAddon.h"
 #include "TimingConstants.h"
-#include "addons/InputStream.h"
+#include "addons/binary-addons/AddonDll.h"
+#include "addons/binary-addons/BinaryAddonBase.h"
 #include "cores/VideoPlayer/DVDClock.h"
-
-CInputStreamAddon::CInputStreamAddon(const CFileItem& fileitem, std::shared_ptr<ADDON::CInputStream> inputStream)
-: CDVDInputStream(DVDSTREAM_TYPE_ADDON, fileitem), m_addon(inputStream)
+#include "cores/VideoPlayer/DVDDemuxers/DVDDemux.h"
+#include "cores/VideoPlayer/DVDDemuxers/DVDDemuxUtils.h"
+#include "filesystem/SpecialProtocol.h"
+#include "utils/StringUtils.h"
+#include "utils/URIUtils.h"
+
+using namespace ADDON;
+
+CInputStreamAddon::CInputStreamAddon(BinaryAddonBasePtr& addonBase, IVideoPlayer* player, const CFileItem& fileitem)
+  : IAddonInstanceHandler(ADDON_INSTANCE_INPUTSTREAM, addonBase),
+    CDVDInputStream(DVDSTREAM_TYPE_ADDON, fileitem),
+    m_player(player)
 {
+  std::string listitemprops = addonBase->Type(ADDON_INPUTSTREAM)->GetValue("@listitemprops").asString();
+  std::string name(addonBase->ID());
+
+  m_fileItemProps = StringUtils::Tokenize(listitemprops, "|");
+  for (auto &key : m_fileItemProps)
```

Comment:
@ksooo you can't trust the compiler because the compiler can't know what you want to do. There are basically three options:
* `for (auto value : values)` which will create a full copy of the object (most often not what you want)
* `for (auto& value : values)` which will create a modifyable reference to the object (similar to `iterator`)
* `for (const auto& value : values` which will create a const reference to the object (similar to `const_iterator`)

This is the same as with regular C++ iterators. You have to tell the compiler whether to use `iterator` or `const_iterator`.